src/commands/publish.js

```diff
@@ -1,6 +1,37 @@
 // @flow
 import type {Args, Opts} from '../types';
+import {PError} from '../utils/errors';
+import * as logger from '../utils/logger';
+import * as locks from '../utils/locks';
+import getUnpublishedPackages from '../utils/getUnpublishedPackages';
+import Project from '../Project';
+import * as npm from '../utils/npm';
+import * as yarn from '../utils/yarn';
+
 
 export default async function publish(args: Args, opts: Opts) {
+  const cwd = process.cwd();
+  const project = await Project.init(cwd);
+  const workspaces = await project.getWorkspaces();
+  const packages = workspaces.map(workspace => workspace.pkg);
+
+  try {
+    await locks.lock(packages);
+    const unpublishedPackages = await getUnpublishedPackages(opts);
+    const isUnpublished = workspace => unpublishedPackages.some(pkg => workspace.pkg.config.name === pkg.name);
+    const unpublishedWorkspaces = workspaces.filter(isUnpublished);
+
+    await Project.runWorkspaceTasks(unpublishedWorkspaces, async workspace => {
+      const {name, version} = workspace.pkg.config;
+      logger.info(`Publishing ${name} at ${version}`);
+
+      await npm.publish(workspace.pkg.config.name, { cwd: workspace.pkg.dir });
+    });
+
+    await locks.unlock(packages);
 
+  } catch (err) {
+    logger.error(err.message);
```

Comment:
We *could* add an
```
await locks.unlock(packages)
```

here, but it would add a lot more noise when you hit the code path where you roll back already

attemptToGetLocks -> somePass, some  fail -> clean up all locks -> rethrow error -> gets caught here -> cleans up again (which is super loud because npm puts three lines of warning on the stderr per package)

Fine for now, but would like to make cleaner

Reply:
The way that I was thinking of handling rollbacks is to do stuff in caught phases:

```js
try {
  await doSomeWork(); // phase 1
  try {
    await doSomeMoreWork(); // phase 2
    try {
      await doFinalWork(); // phase 3
    } catch (err) {
      await undoFinalwork(); // rollback phase 3
      throw err; // rethrow to bubble
    }
  } catch (err) {
    await undoSomeMoreWork(); // rollback phase 2
    throw err; // rethrow to bubble
  }
} catch (err) {
  await undoSomeWork(); // rollback phase 1
  throw err; // rethrow to bubble
}
```

Just as long as we abstract each "phase" into its own function we will avoid this hell, and we can rely on each step rolling itself back and throwing.

I'm hopefully this will work, but I might be wrong– Particularly if we need to rollback in an order other than the reverse order (which I don't think we need to).

Reply:
You could also log in these places as the re-thrown errors bubble up to report what is happening